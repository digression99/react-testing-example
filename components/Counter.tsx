import { useState, useEffect } from 'react'
import {Box, Button, Text} from '@chakra-ui/react'

type CounterOption = {
  min?: number
  max?: number
}

const getErrorMessage = (count: number, min: number, max: number) => {
  if (count >= max) return 'Count is max.'
  if (count <= min) return 'Count is min.'
  return ''
}

const useCounter = (initialCount: number = 0, option: CounterOption) => {
  const [count, setCount]  = useState(initialCount)
  const [error, setError] = useState('')

  const [min, max] = [option.min || -Infinity, option.max || Infinity]

  useEffect(() => {

    setError( getErrorMessage(count, min ,max) )
  }, [count, min, max])

  const increase = () => {
    setCount(Math.min(max, count + 1))
    setError(count >= max ? 'Count is max.' : '')
  }
  const decrease = () => {
    setCount(Math.max(min, count - 1))
    setError(count <= min ? 'Count is min.' : '')
  }
  const reset = () => {
    setCount(initialCount)
    setError('')
  }

  return {
    count,
    error,
    increase,
    decrease,
    reset
  }
}

export const Counter = () => {
  const { count, error, ...methods } = useCounter(0, { min: -5, max: 5 })
  return <PureCounter
    count={count}
    error={error}
    onIncrease={methods.increase}
    onDecrease={methods.decrease}
    onReset={methods.reset}
  />
}

type PureCounterProps = Omit<ReturnType<typeof useCounter>, 'increase' | 'decrease' | 'reset'> & {
  onIncrease: VoidFunction
  onDecrease: VoidFunction
  onReset: VoidFunction
}

export const PureCounter = ({ count, error, onIncrease, onDecrease, onReset}: PureCounterProps) => {
  return (
    <Box>
      <Box>
        <Text>Count : {count}</Text>
      </Box>

      <Box>
        <Text>Controllers</Text>
        <Button onClick={onIncrease}>Increase</Button>
        <Button onClick={onDecrease}>Decrease</Button>
        <Button onClick={onReset}>Reset</Button>
      </Box>

      <Box>
        <Text>{error}</Text>
      </Box>
    </Box>
  )
}







