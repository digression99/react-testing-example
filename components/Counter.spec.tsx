import { render } from '@testing-library/react'
import { PureCounter } from './Counter'

describe('<Counter />', () => {
  test('renders the component correctly', () => {
    const { container: container1 } = render(<PureCounter
      count={0}
      error={''}
      onIncrease={jest.fn()}
      onDecrease={jest.fn()}
      onReset={jest.fn()}
    />)

    expect(container1).toMatchSnapshot()

    const { container: container2 } = render(<PureCounter
      count={100}
      error={'Count is max.'}
      onIncrease={jest.fn()}
      onDecrease={jest.fn()}
      onReset={jest.fn()}
    />)
    expect(container2).toMatchSnapshot()
  })

  test('renders the special component correctly', () => {
    const { container } = render(<PureCounter
      count={30}
      error={'Count is min.'}
      onIncrease={jest.fn()}
      onDecrease={jest.fn()}
      onReset={jest.fn()}
    />)
    expect(container).toMatchSnapshot()
  })
})