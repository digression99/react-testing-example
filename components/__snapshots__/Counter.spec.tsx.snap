// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> renders the component correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Count : 
        0
      </p>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Controllers
      </p>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Increase
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Decrease
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Reset
      </button>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      />
    </div>
  </div>
</div>
`;

exports[`<Counter /> renders the component correctly 2`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Count : 
        100
      </p>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Controllers
      </p>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Increase
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Decrease
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Reset
      </button>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Count is max.
      </p>
    </div>
  </div>
</div>
`;

exports[`<Counter /> renders the special component correctly 1`] = `
<div>
  <div
    class="css-0"
  >
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Count : 
        30
      </p>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Controllers
      </p>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Increase
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Decrease
      </button>
      <button
        class="chakra-button css-0"
        type="button"
      >
        Reset
      </button>
    </div>
    <div
      class="css-0"
    >
      <p
        class="chakra-text css-0"
      >
        Count is min.
      </p>
    </div>
  </div>
</div>
`;
